performance-tests/bench_2/test_controller/ScenarioManager.cpp

```diff
@@ -201,107 +202,254 @@ void ScenarioManager::customize_configs(std::map<std::string, std::string>& work
+    const NodePrototype& nodeproto = scenario_prototype.nodes[i];
+    // Probability that a matched node controller is allocated to a node in this prototype
+    double my_prob = 0.0;
+    if (nodeproto.exclusive && nodeproto.count > 0) {
+      if (nodeproto.count > matched_ncs[i].size()) {
+        std::string msg = "Not enough nodes for node prototype with wildcard '" +
+          std::string(nodeproto.name_wildcard.in()) + "'!";
+        throw std::runtime_error(msg);
+      }
+      my_prob = static_cast<double>(nodeproto.count)/matched_ncs[i].size();
+    }
+
+    // Update the weight for each of the matched node controller
+    for (unsigned j = 0; j < matched_ncs[i].size(); ++j) {
+      const NodeController::NodeId& nc_id = matched_ncs[i][j];
+      if (nc_weights.find(nc_id) != nc_weights.end()) {
+        nc_weights[nc_id].max_prob = std::max(nc_weights[nc_id].max_prob, my_prob);
+        ++nc_weights[nc_id].overlap_count;
+      } else {
+        nc_weights[nc_id].id = nc_id;
+        nc_weights[nc_id].max_prob = my_prob;
+        nc_weights[nc_id].overlap_count = 1;
+      }
+    }
   }
 
-  // Make Sure We Have Enough Nodes
-  if (available_nodes.size() < minimum_nodes) {
-    std::stringstream ss;
-    ss
-      << "At least " << minimum_nodes << " available node" << (minimum_nodes == 1 ? " is " : "s are ")
-      << "required for this scenario.";
-    throw std::runtime_error(ss.str());
+  // Read in worker config files
+  std::map<std::string, std::string> worker_configs;
+  for (unsigned i = 0; i < scenario_prototype.nodes.length(); ++i) {
+    read_protoworker_configs(test_context_, scenario_prototype.nodes[i].workers,
+      worker_configs, debug_alloc);
   }
+  read_protoworker_configs(test_context_, scenario_prototype.any_node, worker_configs, debug_alloc);
 
-  // Start Building the Scenario
-  const unsigned node_count = std::min(static_cast<unsigned>(available_nodes.size()), maximum_nodes);
   AllocatedScenario allocated_scenario;
   allocated_scenario.expected_reports = 0;
   allocated_scenario.timeout = scenario_prototype.timeout;
-  allocated_scenario.configs.length(node_count);
-
-  // Set the Node IDs and Initialize Worker Length
-  auto ani = available_nodes.begin();
-  for (unsigned i = 0; i < node_count; i++) {
-    allocated_scenario.configs[i].node_id = ani->first;
-    allocated_scenario.configs[i].timeout = scenario_prototype.timeout;
-    allocated_scenario.configs[i].workers.length(0);
-    ani++;
-  }
+  allocated_scenario.configs.length(0);
+
+  // Node controllers allocated to exclusive nodes
+  std::set<NodeController::NodeId, OpenDDS::DCPS::GUID_tKeyLessThan> exclusive_ncs;
+
+  // Allocate exclusive node configs first
+  for (unsigned i = 0; i < scenario_prototype.nodes.length(); ++i) {
+    const NodePrototype& nodeproto = scenario_prototype.nodes[i];
+    if (nodeproto.exclusive && nodeproto.count > 0) {
+      // Node controllers that are less likely to be used by some other node prototypes
```

Comment:
Ah, here's the comment I was looking for earlier. Makes sense. Maybe this gets moved (or copied) into the new header.